File: evaluation/tools/build_gold_set.py
#!/usr/bin/env python3
"""Build Gold PR dataset from GitHub pull requests.

Usage:
  python evaluation/tools/build_gold_set.py \
    --input evaluation/data/pr_targets.json \
    --output evaluation/data/gold_pr_set.jsonl

Input format (JSON):
[
  {"repository": "owner/repo", "pr_number": 123},
  {"repository": "owner/repo", "pr_number": 456}
]

Required env:
  GITHUB_TOKEN
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any

from dotenv import load_dotenv


ALLOWED_EXTENSIONS = {
    ".js",
    ".jsx",
    ".ts",
    ".tsx",
    ".vue",
    ".svelte",
    ".css",
    ".scss",
    ".html",
}
SPECIAL_FILES = {
    "package.json",
}


def _api_get(url: str, token: str) -> Any:
    request = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
            "User-Agent": "code-review-agent-eval",
        },
    )
    with urllib.request.urlopen(request, timeout=30) as response:  # noqa: S310
        return json.loads(response.read().decode("utf-8"))


def _is_target_file(path: str) -> bool:
    for filename in SPECIAL_FILES:
        if path.endswith(filename):
            return True
    for ext in ALLOWED_EXTENSIONS:
        if path.endswith(ext):
            return True
    return False


def _normalize_severity(text: str) -> str:
    t = text.lower()
    if any(
        k in t
        for k in [
            "critical",
            "blocker",
            "rce",
            "sql injection",
            "remote code execution",
            "broken access control",
        ]
    ):
        return "critical"
    if any(
        k in t
        for k in [
            "high",
            "xss",
            "security",
            "vulnerability",
            "csrf",
            "ssrf",
            "unsafe deserialization",
            "mass assignment",
        ]
    ):
        return "high"
    if any(
        k in t for k in ["medium", "performance", "slow", "memory", "n+1", "n plus one"]
    ):
        return "medium"
    if any(k in t for k in ["low", "nit", "minor", "style"]):
        return "low"
    return "unknown"


def _normalize_category(text: str) -> str:
    t = text.lower()
    if any(
        k in t
        for k in [
            "xss",
            "csrf",
            "security",
            "token",
            "auth",
            "ssrf",
            "idor",
            "access control",
            "mass assignment",
            "sql injection",
            "unsafe deserialization",
            "cve",
        ]
    ):
        return "security"
    if any(
        k in t
        for k in [
            "slow",
            "performance",
            "render",
            "latency",
            "n+1",
            "index",
            "query plan",
        ]
    ):
        return "performance"
    if any(
        k in t
        for k in [
            "bug",
            "incorrect",
            "wrong",
            "null",
            "undefined",
            "transaction",
            "race condition",
            "consistency",
            "idempotency",
        ]
    ):
        return "correctness"
    if any(k in t for k in ["readability", "refactor", "maintain", "complex"]):
        return "maintainability"
    if any(k in t for k in ["style", "format", "naming", "lint"]):
        return "style"
    return "unknown"


def _extract_line(comment: dict[str, Any]) -> int:
    line = comment.get("line")
    if isinstance(line, int) and line > 0:
        return line
    original_line = comment.get("original_line")
    if isinstance(original_line, int) and original_line > 0:
        return original_line
    return 1


@dataclass
class Target:
    repository: str
    pr_number: int


def load_targets(path: str) -> list[Target]:
    with open(path, encoding="utf-8") as f:
        raw = json.load(f)
    targets: list[Target] = []
    for item in raw:
        repo = item["repository"]
        pr = int(item["pr_number"])
        targets.append(Target(repository=repo, pr_number=pr))
    return targets


def build_gold_item(target: Target, token: str) -> dict[str, Any]:
    owner, repo = target.repository.split("/", maxsplit=1)

    pr_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{target.pr_number}"
    files_url = f"{pr_url}/files?per_page=100"
    review_comments_url = f"{pr_url}/comments?per_page=100"

    pr_data = _api_get(pr_url, token)
    files_data = _api_get(files_url, token)
    comments_data = _api_get(review_comments_url, token)

    file_changes: list[dict[str, Any]] = []
    for file_item in files_data:
        path = file_item.get("filename", "")
        patch = file_item.get("patch")
        if not path or not patch:
            continue
        if _is_target_file(path):
            file_changes.append({"path": path, "patch": patch})

    human_findings: list[dict[str, Any]] = []
    for comment in comments_data:
        body = (comment.get("body") or "").strip()
        path = comment.get("path") or ""
        if not body or not path or not _is_target_file(path):
            continue
        summary = re.sub(r"\s+", " ", body)
        finding = {
            "category": _normalize_category(summary),
            "severity": _normalize_severity(summary),
            "path": path,
            "line": _extract_line(comment),
            "summary": summary,
            "source": comment.get("html_url") or pr_data.get("html_url"),
        }
        human_findings.append(finding)

    labels = [label["name"] for label in pr_data.get("labels", []) if "name" in label]

    return {
        "id": f"{target.repository}#{target.pr_number}",
        "repository": target.repository,
        "pr_number": target.pr_number,
        "title": pr_data.get("title", ""),
        "body": pr_data.get("body") or "",
        "labels": labels,
        "file_changes": file_changes,
        "human_findings": human_findings,
    }


def main() -> int:
    load_dotenv()

    parser = argparse.ArgumentParser(description="Build Gold PR dataset from GitHub")
    parser.add_argument("--input", required=True, help="Path to input target JSON")
    parser.add_argument("--output", required=True, help="Path to output JSONL")
    parser.add_argument(
        "--sleep", type=float, default=0.2, help="Sleep between API calls"
    )
    args = parser.parse_args()

    token = os.getenv("GITHUB_TOKEN")
    if not token:
        print("GITHUB_TOKEN is required", file=sys.stderr)
        return 2

    targets = load_targets(args.input)
    os.makedirs(os.path.dirname(args.output), exist_ok=True)

    count = 0
    with open(args.output, "w", encoding="utf-8") as out:
        for target in targets:
            try:
                item = build_gold_item(target, token)
            except urllib.error.HTTPError as e:
                print(
                    f"[WARN] skip {target.repository}#{target.pr_number}: HTTP {e.code}"
                )
                continue
            except Exception as e:  # noqa: BLE001
                print(f"[WARN] skip {target.repository}#{target.pr_number}: {e}")
                continue

            if not item["file_changes"]:
                print(
                    f"[INFO] no target file changes: {target.repository}#{target.pr_number}"
                )
                continue
            if not item["human_findings"]:
                print(
                    f"[INFO] no review comments: {target.repository}#{target.pr_number}"
                )
                continue

            out.write(json.dumps(item, ensure_ascii=False) + "\n")
            count += 1
            time.sleep(args.sleep)

    print(f"Done. Gold items: {count}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/evaluation/src/build-gold-set.ts
#!/usr/bin/env node
/**
 * Build Gold PR dataset from GitHub pull requests.
 *
 * Usage:
 *   build-gold-set --input evaluation/data/pr_targets.json \
 *     --output evaluation/data/gold_pr_set.jsonl
 *
 * Input format (JSON):
 * [
 *   {"repository": "owner/repo", "pr_number": 123},
 *   {"repository": "owner/repo", "pr_number": 456}
 * ]
 *
 * Required env: GITHUB_TOKEN
 */
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";
import { Command, CommanderError } from "commander";
import { apiGet as defaultApiGet, fetchPrFiles as defaultFetchPrFiles } from "./lib/github-rest.js";
import { writeJsonlAtomic as defaultWriteJsonlAtomic } from "./lib/jsonl.js";
import { getLogger } from "./lib/logging.js";
import { isProductionCodeFile } from "./lib/target-criteria.js";

const logger = getLogger("build_gold_set");

const SEVERITIES: ReadonlySet<string> = new Set(["critical", "high", "medium", "low", "unknown"]);
const IMPACTS: ReadonlySet<string> = new Set([
  "security",
  "correctness",
  "performance",
  "maintainability",
  "unknown",
]);
const PRIORITIES: ReadonlySet<string> = new Set(["high", "medium", "low", "unknown"]);
const STACKS: ReadonlySet<string> = new Set(["react", "vue", "angular", "svelte"]);

export interface Target {
  repository: string;
  pr_number: number;
  stack: string;
  severity: string;
  impact: string;
  priority: string;
}

export interface FileChange {
  path: string;
  patch: string;
}

export interface HumanFinding {
  category: string;
  severity: string;
  impact: string;
  priority: string;
  path: string;
  line: number;
  summary: string;
  source: string | undefined;
}

export interface GoldItem {
  id: string;
  repository: string;
  pr_number: number;
  stack: string;
  title: string;
  body: string;
  labels: string[];
  file_changes: FileChange[];
  human_findings: HumanFinding[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function normalizeAxis(value: unknown, choices: ReadonlySet<string>): string {
  if (typeof value !== "string") {
    return "unknown";
  }
  const normalized = value.trim().toLowerCase();
  return choices.has(normalized) ? normalized : "unknown";
}

/** Load and validate PR targets, defaulting missing/invalid axes to "unknown". */
export function loadTargets(path: string): Target[] {
  const raw = JSON.parse(readFileSync(path, "utf-8"));
  if (!Array.isArray(raw)) {
    throw new Error(`input is not a JSON array: ${path}`);
  }
  return raw.map((item: unknown, index: number) => {
    if (!isRecord(item)) {
      throw new Error(`invalid target at [${index}]`);
    }
    const repository = String(item.repository);
    const prNumber = Number(item.pr_number);
    const stack = item.stack;
    if (typeof stack !== "string" || !STACKS.has(stack)) {
      const allowed = [...STACKS].sort().join(", ");
      throw new Error(
        `invalid target at [${index}]: stack=${JSON.stringify(stack)}; expected one of: ${allowed}`,
      );
    }
    return {
      repository,
      pr_number: prNumber,
      stack,
      severity: normalizeAxis(item.severity, SEVERITIES),
      impact: normalizeAxis(item.impact, IMPACTS),
      priority: normalizeAxis(item.priority, PRIORITIES),
    };
  });
}

const CATEGORY_KEYWORDS: [string, string[]][] = [
  [
    "security",
    [
      "xss",
      "csrf",
      "security",
      "token",
      "auth",
      "ssrf",
      "idor",
      "access control",
      "mass assignment",
      "sql injection",
      "unsafe deserialization",
      "cve",
    ],
  ],
  ["performance", ["slow", "performance", "render", "latency", "n+1", "index", "query plan"]],
  [
    "correctness",
    [
      "bug",
      "incorrect",
      "wrong",
      "null",
      "undefined",
      "transaction",
      "race condition",
      "consistency",
      "idempotency",
    ],
  ],
  ["maintainability", ["readability", "refactor", "maintain", "complex"]],
  ["style", ["style", "format", "naming", "lint"]],
];

/** Classify a finding summary into a category by keyword, in priority order. */
export function normalizeCategory(text: string): string {
  const lowered = text.toLowerCase();
  for (const [category, keywords] of CATEGORY_KEYWORDS) {
    if (keywords.some((keyword) => lowered.includes(keyword))) {
      return category;
    }
  }
  return "unknown";
}

function isPositiveInt(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value > 0;
}

function extractLine(comment: Record<string, unknown>): number {
  if (isPositiveInt(comment.line)) {
    return comment.line;
  }
  if (isPositiveInt(comment.original_line)) {
    return comment.original_line;
  }
  return 1;
}

export type ApiGet = (url: string, token: string) => Promise<unknown>;
export type FetchPrFiles = (
  owner: string,
  repo: string,
  prNumber: number,
  token: string,
) => Promise<FileChange[]>;

const REVIEW_COMMENTS_PER_PAGE = 100;
const REVIEW_COMMENTS_MAX_PAGES = 100;

async function fetchReviewComments(
  owner: string,
  repo: string,
  prNumber: number,
  token: string,
  apiGet: ApiGet,
): Promise<Record<string, unknown>[]> {
  const comments: Record<string, unknown>[] = [];
  for (let page = 1; page <= REVIEW_COMMENTS_MAX_PAGES; page += 1) {
    const url =
      `https://api.github.com/repos/${owner}/${repo}/pulls/${prNumber}/comments` +
      `?per_page=${REVIEW_COMMENTS_PER_PAGE}&page=${page}`;
    const pageData = await apiGet(url, token);
    if (!Array.isArray(pageData)) {
      throw new TypeError("GitHub PR review comments response must be an array");
    }
    comments.push(...(pageData as Record<string, unknown>[]));
    if (pageData.length < REVIEW_COMMENTS_PER_PAGE) {
      return comments;
    }
  }
  throw new Error(
    `GitHub PR review comments pagination exceeded ${REVIEW_COMMENTS_MAX_PAGES} pages`,
  );
}

export interface BuildGoldItemDeps {
  apiGet?: ApiGet;
  fetchPrFiles?: FetchPrFiles;
}

/** Split "owner/repo" into its two path segments (repo may itself contain no further slash). */
function splitRepository(repository: string): [string, string] {
  const idx = repository.indexOf("/");
  if (idx === -1) {
    throw new Error(`invalid repository: ${repository}`);
  }
  return [repository.slice(0, idx), repository.slice(idx + 1)];
}

export async function buildGoldItem(
  target: Target,
  token: string,
  deps: BuildGoldItemDeps = {},
): Promise<GoldItem> {
  const apiGet = deps.apiGet ?? defaultApiGet;
  const fetchPrFiles = deps.fetchPrFiles ?? defaultFetchPrFiles;
  const [owner, repo] = splitRepository(target.repository);

  const prUrl = `https://api.github.com/repos/${owner}/${repo}/pulls/${target.pr_number}`;
  const prData = (await apiGet(prUrl, token)) as Record<string, unknown>;
  const files = await fetchPrFiles(owner, repo, target.pr_number, token);
  const comments = await fetchReviewComments(owner, repo, target.pr_number, token, apiGet);

  const fileChanges: FileChange[] = files
    .filter((file) => isProductionCodeFile(file.path))
    .map((file) => ({ path: file.path, patch: file.patch }));

  const humanFindings: HumanFinding[] = [];
  for (const comment of comments) {
    const body = typeof comment.body === "string" ? comment.body.trim() : "";
    const path = typeof comment.path === "string" ? comment.path : "";
    if (!body || !path || !isProductionCodeFile(path)) {
      continue;
    }
    const summary = body.replace(/\s+/g, " ");
    humanFindings.push({
      category: normalizeCategory(summary),
      severity: target.severity,
      impact: target.impact,
      priority: target.priority,
      path,
      line: extractLine(comment),
      summary,
      source: (comment.html_url as string | undefined) ?? (prData.html_url as string | undefined),
    });
  }

  const labels = Array.isArray(prData.labels)
    ? (prData.labels as unknown[])
        .filter((label): label is Record<string, unknown> => isRecord(label) && "name" in label)
        .map((label) => String(label.name))
    : [];

  return {
    id: `${target.repository}#${target.pr_number}`,
    repository: target.repository,
    pr_number: target.pr_number,
    stack: target.stack,
    title: typeof prData.title === "string" ? prData.title : "",
    body: typeof prData.body === "string" ? prData.body : "",
    labels,
    file_changes: fileChanges,
    human_findings: humanFindings,
  };
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function defaultSleep(seconds: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, seconds * 1000);
  });
}

export interface RunDeps {
  apiGet?: ApiGet;
  fetchPrFiles?: FetchPrFiles;
  writeJsonlAtomic?: typeof defaultWriteJsonlAtomic;
  sleep?: (seconds: number) => Promise<void>;
  env?: NodeJS.ProcessEnv;
}

interface ParsedOptions {
  input: string;
  output: string;
  sleep: number;
}

export async function run(argv: string[], deps: RunDeps = {}): Promise<number> {
  const env = deps.env ?? process.env;
  const writeJsonlAtomic = deps.writeJsonlAtomic ?? defaultWriteJsonlAtomic;
  const sleep = deps.sleep ?? defaultSleep;

  const program = new Command();
  program
    .name("build-gold-set")
    .description("Build Gold PR dataset from GitHub")
    .requiredOption("--input <path>", "Path to input target JSON")
    .requiredOption("--output <path>", "Path to output JSONL")
    .option("--sleep <seconds>", "Sleep between API calls", (v) => Number.parseFloat(v), 0.2)
    .allowExcessArguments(false)
    .exitOverride();

  try {
    program.parse(argv, { from: "user" });
  } catch (error) {
    if (error instanceof CommanderError) {
      return error.exitCode === 0 ? 0 : 2;
    }
    throw error;
  }

  const options = program.opts<ParsedOptions>();

  const token = env.GITHUB_TOKEN;
  if (!token) {
    logger.error("GITHUB_TOKEN is required");
    return 2;
  }

  const targets = loadTargets(options.input);

  const items: GoldItem[] = [];
  for (const target of targets) {
    let item: GoldItem;
    try {
      item = await buildGoldItem(target, token, deps);
    } catch (error) {
      logger.warn(`skip ${target.repository}#${target.pr_number}: ${errorMessage(error)}`);
      continue;
    }

    if (item.file_changes.length === 0) {
      logger.info(`no target file changes: ${target.repository}#${target.pr_number}`);
      continue;
    }
    if (item.human_findings.length === 0) {
      logger.info(`no review comments: ${target.repository}#${target.pr_number}`);
      continue;
    }

    items.push(item);
    await sleep(options.sleep);
  }

  await writeJsonlAtomic(options.output, items);
  logger.info(`Done. Gold items: ${items.length}`);
  return 0;
}

export function isDirectExecution(
  metaUrl: string = import.meta.url,
  entrypoint: string | undefined = process.argv[1],
): boolean {
  return entrypoint !== undefined && metaUrl === pathToFileURL(entrypoint).href;
}

export async function main(): Promise<number> {
  return run(process.argv.slice(2));
}

if (isDirectExecution()) {
  main()
    .then((code) => {
      process.exitCode = code;
    })
    .catch((err) => {
      console.error(err);
      process.exitCode = 1;
    });
}
